### ISSUE TITLE
Telemetry vs Property data

### ISSUE THREAD

**Issue by MLopezJ** [MLopezJ/asset-tracker-cloud-coiote-azure-converter-js#18]
# TL;DR

The different between those 2 data type is that "Telemetry" are sent via device-to-cloud and "Property" via Device Twin. Telemetry data is stored and are sent to the Azure event bus, which can be processed in a more convenient way than periodic refreshing of the device twin.

There is not functionality related to "telemetry" data type. 

In the current context, there is no real benefit in define values as "telemetry"

# Issue discussion
[Expected Input](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js#expected-input) data relies in the assumption that [defining data as "property" on Coiote](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/issues/8#issuecomment-1428210012) will remove the issue rewarding [not expected values in Device Twin](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/issues/7)

It is important to understand the implications of it. 

## Questions

Why does marking resource as `Property` produce the value in the shadow but why does it not when marking it as `Telemetry`?

**Comment by MLopezJ** [comment]
Question asked to AVSystem: 

> # Implication of Telemetry and Property in objects definition
> ## Context
> 
> - I followed the next tutorial: https://iotdevzone.avsystem.com/docs/Demo_Projects/Tracking_tutorial/
> - I stopped before start with Part 4 because I don't need the Power BI integration
> - As result of the tutorial I’m having an IoT Device Hub in Azure. 
> - The Device twin of the IoT Hub contains the data showed on the Coiote Dashboard
> 
> ## Problem
> 
> The device twin contains  the following values: { 'pmin': 600, 'pmax': 3600 } for the next objects: 
> 
> - Object: 6 Prop: 0
> - Object: 6 Prop: 1
> - Object: 3303 Prop: 5700
> - Object: 3304 Prop: 5700
> - Object: 3315 Prop: 5700
> 
> Value returned was not expected because on the Coiote Dashboard the value are data that make sense with its LWM2M definition. 
> 
> Device Twin result: https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/tree/128038853cfbe9bc45ce676d776de64733528a5a#expected-input
> 
> ## Solution
> 
> I made a new integration and I change definition of values from “Telemetry” to “Property”. The opposite that is requested here: [https://iotdevzone.avsystem.com/docs/Demo_Projects/Tracking_tutorial/#:~:text=Set the CAPABILITY TYPE to Telemetry for the resources%3A](https://iotdevzone.avsystem.com/docs/Demo_Projects/Tracking_tutorial/#:~:text=Set%20the%20CAPABILITY%20TYPE%20to%20Telemetry%20for%20the%20resources%3A)
> 
> ## Result
> 
> { 'pmin': 600, 'pmax': 3600 } format is not provided anymore as value in the device twin, new values make sense with its definition and with  reported on Coiote Dashboard
> 
> Device Twin result:  https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/tree/e92b83b864e6a85b902a9b13e95684ab44441c51#expected-input
> 
> ## Questions
> 
> Why does marking resource as Property produce the value in the shadow but why does it not when marking it as Telemetry?
> 
> What is the implication of that change?

Status: waiting for response

**Comment by MLopezJ** [comment]
Pause the effort to [document the Integration steps](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/issues/14) until resolve this issue

**Comment by MLopezJ** [comment]
> Why does marking resource as `Property` produce the value in the shadow but why does it not when marking it as `Telemetry`?

Because `Telemetry` data use another channel to report data. Also, historical values from `Telemetry` data are stored.

## differences 
| | Telemetry | Property|
| --- | --- | ---|
| **Chanel**|  device-to-cloud messaging | Device twin |
| **Historical data storage** | Yes| No

**Comment by MLopezJ** [comment]
# Which values from nRF Asset Tracker need to have the historical data record? 

- Location
- Battery
- Roaming

# What is "device-to-cloud messaging mechanism" ? How it works?

**Comment by MLopezJ** [comment]
After the answer from AVSystem there is still the question about what is the implication of changing the setting of the objects from `Telemetry` to `Property`. It is only the lose in the capability of store historical data?

This is the answer from AVSystme:
```
Hello,

This is because Telemetry uses a device-to-cloud messaging mechanism ([https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-d2c) ](https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-d2c))
instead of device twin. That's the difference between Telemetry and Property. Also, for 
resources that are marked as Property, only the current value will be stored, and for Telemetry, 
historical values will also be stored.

You can read more about LwM2M mappings 
for Azure IoT Hub here: https://iotdevzone.avsystem.com/docs/Cloud_integrations/Azure_IoT/Concepts/LwM2M_mappings_Hub/
```

The first link provided does not work. 

I will read about `device-to-cloud messaging mechanism` to see if I can get others implication of that change. 

Also, it is important to understand if historical data is needed in the web app. I think the answer is yes but I dont know how is  treated the historical data there. I need more details about it. 

Another situation is that if historical data is needed and they will be in another channel, **how both channels will be integrated in [one output](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js#expected-output)?** Because nRF Asset Tracker is requesting [Location, Temperature and Humidity ](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/issues/7#issuecomment-1424426763) as [part of it input](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/saga/docs/cloud-protocol/state.reported.schema.json)

**Comment by MLopezJ** [comment]
> Which values from nRF Asset Tracker need to have the historical data record?

In the [documentation](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/886e2354563d02fcd7eacdeeb6675670f9b55434/docs/app/Index.rst) there is said:

```
The web application offers the following features:

- Viewing of asset trackers
  - Current and historical device data, for example
    - GNSS location
    - Approximate location using :ref:`nRF Cloud Location Services <app-cellgeolocation>` and other third-party providers
    - Battery voltage
    - Temperature
```
And at the [code level](https://github.com/NordicSemiconductor/asset-tracker-cloud-app-js/blob/aefb903eb08e0a658f3cb35fa06af8d7304c4bb1/src/azure/CatsMap/Loader.tsx), the `queryHistoricalDeviceData` method is sending a query as param to return data related to GNSS, which is[ related to Location](https://github.com/MLopezJ/nRF-Asset-Tracker-through-Coiote-flow/blob/saga/docs/gnss.md). 

Location is [one of the values](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/issues/7#issuecomment-1424426763) settled as `telemetry`

So the Location object need to be store somewhere in order to the web app work properly.

**Comment by MLopezJ** [comment]
### [Historical data](https://github.com/NordicSemiconductor/asset-tracker-cloud-app-js/blob/aefb903eb08e0a658f3cb35fa06af8d7304c4bb1/src/azure/CatsMap/Loader.tsx#L16) for Location object is needed in the web app. 

There are indications that historical data is also needed for:

- [Battery data ](https://github.com/NordicSemiconductor/asset-tracker-cloud-app-js/commit/8e86731a13d37df0319fb1c60b794a43348f4190)
- [roaming data ](https://github.com/NordicSemiconductor/asset-tracker-cloud-app-js/issues/318)

**Comment by MLopezJ** [comment]
> So the Location object need to be store somewhere in order to the web app work properly.

~Where does the Location data will be storage?~

~And how the app is [storing hisotrical values right now](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/3c028b614159053fc1842590f49ee555dc2c5df8/docs/azure/Index.rst)?~

**Comment by MLopezJ** [comment]
> Where does the Location data will be storage?

It depends on where Telemetry data is stored

> historical values from `Telemetry` data are stored.

# Where are Telemetry data stored??

> And how the app is [storing hisotrical values right now](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/3c028b614159053fc1842590f49ee555dc2c5df8/docs/azure/Index.rst)?

The nRF-Azure backend is using Cosmos DB to storage historical data.

**Comment by MLopezJ** [comment]
# Telemetry data in Coiote

| ID  |  Object  | Prop | Name |  LwM2M Def |
|---|---|---|---|---|
|  6 |  Location | 0 | Latitude |   [link](https://github.com/OpenMobileAlliance/lwm2m-registry/blob/prod/6.xml) | 
|  6 |  Location | 1 | Longitude |   [link](https://github.com/OpenMobileAlliance/lwm2m-registry/blob/prod/6.xml) | 
| 3303  | Temperature  | 5700  | Sensor Value |   [link](https://github.com/OpenMobileAlliance/lwm2m-registry/blob/prod/3303.xml)  | 
| 3304  | Humidity  | 5700  | Sensor Value |   [link](https://github.com/OpenMobileAlliance/lwm2m-registry/blob/prod/3304.xml)  | 
| 3315  | Barometer | 5700 | Sensor Value |    [link](https://github.com/OpenMobileAlliance/lwm2m-registry/blob/prod/3315.xml) | 


# Telemetry data in nRF Asset Tracker
All. Everything in nRF asset tracker is telemetry data. And everything should be storage as historical data except A-GPS and P-GPS.

**Comment by MLopezJ** [comment]
All the values requested in the input data by nRF Asset Tracker should be store to allows the correct functionality of the app. On other hand, defining values as "Telemetry" in Coiote, stores the data but also split up it in 2 different channels and only stores the data defined as "Telemetry". So, the store of the other data should be done manually and also the reconciliation of the two channels as well. 

In that sense, defining "Telemetry" data to not brings any beneficent. And if the only implication of it is just the storage capability, there is no really any incentive to make this division. 

It is desired to double check with Coiote if any functionality is related to the definition of "telemetry" data. If answer is no, all values will be defined as "property" in Coiote.

**Comment by MLopezJ** [comment]
Question asked to AVSystem:

> Thanks for the answer.
> 
> Before close this ticket I would like to double check something: is there any functionality related to "telemetry" data?
> 
> I would like to understand the implication of defining objects as "property" instead of "telemetry" and so far I only saw the lose on the storage capability and the fact that that data will be in a different channel.
> 
> I'm trying to be very explicit on it because I'm planning to define Temperature, Humidity, Accelerometer, Barometer and LED color as "property" even when here those are recommended as "telemetry". https://iotdevzone.avsystem.com/docs/Demo_Projects/Tracking_tutorial/#:~:text=Set%20the%20CAPABILITY%20TYPE%20to%20Telemetry%20for%20the%20resources%3A

> Note: I'm not following completely the tutorial. For what I need, I just need the data on the device twin, so I stopped here https://iotdevzone.avsystem.com/docs/Demo_Projects/Tracking_tutorial/#part-4-connect-microsoft-azure-to-power-bi .
> 
> What I mean about functionalities related to telemetry data is in the context of Coiote functionality and not in the context of just the tutorial.

**Comment by MLopezJ** [comment]
All values are going to be defined as "Property". No functionalities are related to "Telemetry" data and in the current context, that specification is not needed. 

Complete Answer

> Hello,
> 
> Resources marked as "telemetry" are sent via a slightly different communication channel (device-to-cloud) and are sent to the Azure event bus, which can be processed in a more convenient way than periodic refreshing of the device twin. Also, "telemetry" is more convenient to visualize data, since historical values will also be stored. But if you only want to get data from the device twin you can mark these resources as "properties".
> 
> I hope that clears things up. Please let me know if you have additional questions.

**Comment by MLopezJ** [comment]
# Implications of Property data over Telemetry data 

> The different between those 2 data type is that "Telemetry" are sent via device-to-cloud and "Property" via Device Twin. Telemetry data is stored and are sent to the Azure event bus, which can be processed in a more convenient way than periodic refreshing of the device twin.
>
> There is not functionality related to "telemetry" data type.
>
> In the current context, there is no real benefit in define values as "telemetry"

## Questions:
- What is the current context? 
- Where the Telemetry data is stored?
- How to access historical telemetry data? 
- For how long does the historical telemetry data is stored?
- How can I query Telemetry historical data? For example average Temperature in the last 2 months. 
- Does telemetry has any different in speed in order with property?

**Comment by MLopezJ** [comment]
# Implications of `Property` data over `Telemetry` data 

> The different between those 2 data type is that "Telemetry" are sent via device-to-cloud and "Property" via Device Twin. Telemetry data is stored and are sent to the Azure event bus, which can be processed in a more convenient way than periodic refreshing of the device twin.
>
> There is not functionality related to "telemetry" data type.
>
> In the current context, there is no real benefit in define values as "telemetry"

## Questions:
- What is the current context? 
- Where the `Telemetry data` is stored?
- How to access historical `telemetry` data? 
- For how long does the `historical telemetry data` is stored?
- How can I query `Telemetry historical data`? For example average Temperature in the last 2 months. 
- Is `Telemetry` faster than 'Property`?

**Comment by MLopezJ** [comment]
# Current context

`nRF Asset Tracker` is a web application. As part of its requirements to work properly, this app is requesting some data as the input and it must be stored to be able to do queries. In the context of this project, the data must come from `Coiote`.

# Coiote - Azure
`Coiote` has an integration with `Azure` which will make possible to provide the data to `nRF Asset Tracker`. To set up the integration it is required to define a `data schema` and describe the objects as `telemetry` or `property` data.

Main differences between data types are described in the following table 

| | `Telemetry` | `Property` |
| --- | --- | ---|
| **Chanel**|  device-to-cloud messages | Device twin |
| **Historical data storage** | Yes| No

# Implications with AVSystem's answer

> ## First answer
> This is because Telemetry uses a device-to-cloud messaging mechanism instead of device twin. That's the difference between Telemetry and Property. Also, for resources that are marked as Property, only the current value will be stored, and for Telemetry, historical values will also be stored.
> ## Second answer
> Resources marked as "telemetry" are sent via a slightly different communication channel (device-to-cloud) and are sent to the Azure event bus, which can be processed in a more convenient way than periodic refreshing of the device twin. Also, "telemetry" is more convenient to visualize data, since historical values will also be stored. But if you only want to get data from the device twin you can mark these resources as "properties".

- `Telemetry` data is easier to process
- `Telemetry` historical data is stored 


# Implications with Azure definitions

Behind the scenes, `telemetry` is `device-to-cloud` mechanism and `property` is `device twin` mechanism. Those mechanism are defined by Azure to send information from an application to IoT Hub. More info [here](https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-d2c-guidance)

| Factor | Device-to-Cloud | Device Twin |
| --- | --- | --- |
| Storage  | Temporarily stored by IoT Hub, up to seven days | Stored by IoT Hub in the device twin |
| retrieval | Only sequential reading | Retrievable using the [IoT Hub query language](https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-query-language) |
| Size | Up to 256-KB messages | Max reported properties size is 32 KB. |
| Frequency | [High](https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-quotas-throttling) |  Medium |
| Protocol | Available on all protocols |  MQTT and AMQP |

# Questions to AVSystem

> 
> Property data vrs Telemetry data
> 
> Previously I made a question related to this. It is here: https://helpcenter.avsystem.com/plugins/servlet/desk/portal/17/CDM-243
> 
> Now I have some new questions. But before I want to exposes my main assumption. 
> 
> Azure provide 3 mechanism to send information from an application to IoT Hub (https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-d2c-guidance): 
> 
> 1. Device-to-Cloud
> 2. Device Twin
> 3. File Uploads
> 
> I'm assuming that Device-to-cloud is the mechanism behind the scenes for 'telemetry data' and device twin the one for 'property data'. Am I correct on it?
> 
> If so, I have the following questions:
> 
> Before I was told that historical telemetry data is stored but Azure said that Device-to-Cloud is temporarily stored (up to seven days). So... how much does the storage of Telemetry data last? 7 days as well? Or Coiote stores this in a different place? If so, where this data is store? And how can I access it? If I can access it, am I able to do queries as for example average Temperature in the last 2 months? 
> 
> On other hand, the max reported size for device twin object is 32KB. Is Coiote restringing payload bigger than 32KB? If yes, how it is communicated to the device? If not, how it is reported to Azure? It is possible to have logs related to it? 

> ### Shorter version 
> 
> Title: Property data vrs Telemetry data
> 
> Previously I made a question related to this. It is here: https://helpcenter.avsystem.com/plugins/servlet/desk/portal/17/CDM-243
> 
> As part of the last ticket now I'm assuming that Device-to-cloud is the mechanism behind the scenes for 'telemetry data' and device twin the one for 'property data'. Am I correct on it?
> 
> If so, I have the following questions:
> 
> Before I was told that historical telemetry data is stored but Azure said that Device-to-Cloud is temporarily stored (up to seven days https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-d2c-guidance#:~:text=by%20IoT%20Hub%2C-,up%20to%20seven%20days,-.%20Only%20sequential%20reading). So... 
> 
> 1. how much does the storage of Telemetry data last? 7 days as well? Or Coiote stores this in a different place? 
> 1.1 If so, where this data is store? And how can I access it? 
> 1.1.1 If I can access it, am I able to do queries as for example average Temperature in the last 2 months?
> 
> On other hand, the max reported size for device twin object is 32KB. https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-d2c-guidance#:~:text=Maximum%20reported%20properties%20size%20is%2032%20KB.
> 
> 2. Is Coiote restringing payload bigger than 32KB? 
> 2.1 If yes, how it is communicated to the device? 
> 2.2 If not, how it is reported to Azure? It is possible to have logs related to it?
> 
> Thanks, 

Question is not going to be asked to AV System because after a discussion with Markus I got the answers I was looking for.

Answer 1: Coiote stores data on their own, independently of the integration with Azure. It means telemetry data will be available 7 day in device-to-cloud from IoT Hub in Azure and 30 days available in Coiote's database. 

Answer 1.1: Data is stored in a database by Coiote, it can be accessed but I dont know how. But is is not relevant for this use case. 

Answer 1.1.1: Is not possible to do "complex" queries as described in the question

Answer 2: I dont know but payload from device is smaller than 32KB. So it is difficult than we end in a situation where the origin of the issue is because of device twin max size is exceeded


# Questions to myself
1. Is the size of requested input in `nRF Asset Tracker` smaller than 32 KB?
> yes, it is 
2. Is the frequency "medium" enough for `nRF Asset Tracker`. What does "medium" really mean?

# Conclusions 

- In the `nRF Asset Tracker` app, all the data that is requested as input must be stored (except A-GPS and P-GPS). 
- There is not any functionality specifically related to `telemetry` data or `property` data. In other words, both types of data have same functionalities/features/methods related
- /// working on it ///


https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-d2c-guidance

**Comment by MLopezJ** [comment]
> 2\. What does "medium" really mean?

In [this articule](https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-d2c-guidance#:~:text=and%20throttling.-,Medium,-.%20For%20more%20information), Azure uses "frequency" as a factor to compare "Device-to-cloud" and "Device Twin" messages. In that context, "frequency" means how often messages are sent from a device to Azure IoT Hub. The frequency can affect performance and cost of the solution, so it is important to undertand what "medium" means.

The articule categorize the data transmission in 3 differernt levels (low, medium and high) but do not specified the rate for those values and link to this [site](https://learn.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-quotas-throttling) for more information about "quotas amd throttling"

## IoT hub quotas and throttling

Both terms, quotas and throttling, refers to the mechanism Azure IoT Hub has to limit usage of the service. 

I can not confirm a correlation between the `frequency` categorizes and the `tier operations` rank presented in this document. But there is described some chatacteristic about Device Twin than are usefull in the intentation of understand the implications of Device Twin specifications. 

Tier Operation rank:
* Free
* B1 and S1
* B2 and S2
* B3 and S3

I can not neither to confirm in which of the last rank option the current solution is because what the config overview said about it is that there is used the  `standard` tier, without specify wich type it is. The daily message limit is 400,000

![Screenshot from 2023-04-04 13-55-58](https://user-images.githubusercontent.com/8351865/229784144-02caa7dd-7d32-4c05-a541-162de8d4c806.png)


Due to it is not possible to know what is the tier type, it is taken as a reference the most basic tier option. the characteristics are: the device twin can read 100 request per second, updated 50 request per second and do 10 jobs operations per second. The max size of the `desired` and `reported` properties are 32KB each (64KB in total). 

Considering the use case context, 1 device pushing data every 1 second to IoT Hub, the metrics presented for the basic standard tier are good enough for doing the desired tasks. 

As a conclusion, it is not possible to know what "frequency medium" means but with the most basic tier option, the device twin is able to do the expected tasks.
